File: src/main/java/org/oscarehr/integration/myhealthaccess/model/MHAPatient.java
/**
 * Copyright (c) 2012-2018. CloudPractice Inc. All Rights Reserved.
 * This software is published under the GPL GNU General Public License.
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * This software was written for
 * CloudPractice Inc.
 * Victoria, British Columbia
 * Canada
 */

package org.oscarehr.integration.myhealthaccess.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.oscarehr.integration.myhealthaccess.dto.PatientTo1;
import org.springframework.beans.BeanUtils;
import oscar.util.ConversionUtils;

import java.time.LocalDate;
import java.util.Optional;

@NoArgsConstructor
public class MHAPatient
{
	private String id;
	private String email;
	private String firstName;
	private String middleName;
	private String lastName;
	private LocalDate birthDate;
	private String sex;

	private String healthNumber;
	private String healthNumberVersion;
	private PROVINCE_CODES healthCareProvinceCode;

	private String address1;
	private String address2;
	private String city;
	private PROVINCE_CODES addressProvinceCode;
	private String postalCode;
	private String countryCode;

	private String homePhone;
	private String cellPhone;
	private String workPhone;
	private String primaryFax;

	@Setter
	@Getter
	private Boolean hasVoipToken;
	private LINK_STATUS linkStatus;
	@Getter
	@Setter
	private boolean canMessage;
	@Setter
	private String demographicNo;

	public enum PROVINCE_CODES
	{
		UNKNOWN,
		NL,
		PE,
		NS,
		NB,
		QC,
		ON,
		MB,
		SK,
		AB,
		BC,
		YT,
		NT,
		NU
	}

	public enum LINK_STATUS
	{
		NO_LINK,
		PATIENT_REJECTED,
		CLINIC_REJECTED,
		PENDING_CLINIC_CONFIRM,
		PENDING_PATIENT_CONFIRM,
		CONFIRMED,
		VERIFIED;

		@JsonCreator
		public static LINK_STATUS fromString(String str)
		{
			// Old MHA server will send status 'active' instead of 'confirmed' or 'verified'
			// Convert the value. TODO Can be removed once MHA master branch includes MHA-2069.
			if (str.equals("active"))
			{
				str = "confirmed";
			}

			return LINK_STATUS.valueOf(str.toUpperCase());
		}
	}

	public static boolean isValidProvinceCode(String provinceCode)
	{
		for (PROVINCE_CODES validProvinceCode : PROVINCE_CODES.values())
		{
			if (validProvinceCode.name().equals(provinceCode))
			{
				return true;
			}
		}

		return false;
	}

	public static PROVINCE_CODES stringToProvinceCode(String provinceCode)
	{
		if (ConversionUtils.hasContent(provinceCode))
		{
			return PROVINCE_CODES.valueOf(provinceCode);
		}
		return PROVINCE_CODES.UNKNOWN;
	}

	public MHAPatient(PatientTo1 patientTo1)
	{
		BeanUtils.copyProperties(patientTo1, this, "addressProvinceCode", "healthCareProvinceCode", "linkStatus");
		this.healthCareProvinceCode = stringToProvinceCode(patientTo1.getHealthCareProvinceCode());
		this.addressProvinceCode = stringToProvinceCode(patientTo1.getAddressProvinceCode());
		this.linkStatus = patientTo1.getLinkStatus();
	}

	public String getId()
	{
		return id;
	}

	public void setId(String id)
	{
		this.id = id;
	}

	public String getEmail()
	{
		return email;
	}

	public void setEmail(String email)
	{
		this.email = email;
	}

	public String getFirstName()
	{
		return firstName;
	}

	public void setFirstName(String firstName)
	{
		this.firstName = firstName;
	}

	public String getMiddleName()
	{
		return middleName;
	}

	public void setMiddleName(String middleName)
	{
		this.middleName = middleName;
	}

	public String getLastName()
	{
		return lastName;
	}

	public void setLastName(String lastName)
	{
		this.lastName = lastName;
	}

	public LocalDate getBirthDate()
	{
		return birthDate;
	}

	public void setBirthDate(LocalDate birthDate)
	{
		this.birthDate = birthDate;
	}

	public String getSex()
	{
		return sex;
	}

	public void setSex(String sex)
	{
		this.sex = sex;
	}

	public String getHealthNumber()
	{
		return healthNumber;
	}

	public void setHealthNumber(String healthNumber)
	{
		this.healthNumber = healthNumber;
	}

	public String getHealthNumberVersion()
	{
		return healthNumberVersion;
	}

	public void setHealthNumberVersion(String healthNumberVersion)
	{
		this.healthNumberVersion = healthNumberVersion;
	}

	public PROVINCE_CODES getHealthCareProvinceCode()
	{
		return healthCareProvinceCode;
	}

	public void setHealthCareProvinceCode(PROVINCE_CODES healthCareProvinceCode)
	{
		this.healthCareProvinceCode = healthCareProvinceCode;
	}

	public String getAddress1()
	{
		return address1;
	}

	public void setAddress1(String address1)
	{
		this.address1 = address1;
	}

	public String getAddress2()
	{
		return address2;
	}

	public void setAddress2(String address2)
	{
		this.address2 = address2;
	}

	public String getCity()
	{
		return city;
	}

	public void setCity(String city)
	{
		this.city = city;
	}

	public PROVINCE_CODES getAddressProvinceCode()
	{
		return addressProvinceCode;
	}

	public void setAddressProvinceCode(PROVINCE_CODES addressProvinceCode)
	{
		this.addressProvinceCode = addressProvinceCode;
	}

	public String getPostalCode()
	{
		return postalCode;
	}

	public void setPostalCode(String postalCode)
	{
		this.postalCode = postalCode;
	}

	public String getCountryCode()
	{
		return countryCode;
	}

	public void setCountryCode(String countryCode)
	{
		this.countryCode = countryCode;
	}

	public String getHomePhone()
	{
		return homePhone;
	}

	public void setHomePhone(String homePhone)
	{
		this.homePhone = homePhone;
	}

	public String getCellPhone()
	{
		return cellPhone;
	}

	public void setCellPhone(String cellPhone)
	{
		this.cellPhone = cellPhone;
	}

	public String getWorkPhone()
	{
		return workPhone;
	}

	public void setWorkPhone(String workPhone)
	{
		this.workPhone = workPhone;
	}

	public String getPrimaryFax()
	{
		return primaryFax;
	}

	public void setPrimaryFax(String primaryFax)
	{
		this.primaryFax = primaryFax;
	}

	public LINK_STATUS getLinkStatus()
	{
		return linkStatus;
	}

	public void setLinkStatus(LINK_STATUS linkStatus)
	{
		this.linkStatus = linkStatus;
	}

	public Optional<String> getDemographicNo()
	{
		return Optional.ofNullable(this.demographicNo);
	}
}

File: src/main/frontend/src/lib/integration/myhealthaccess/model/MhaPatient.ts
import {LinkStatus} from "./LinkStatus";
import {Province} from "../../../constants/Province";
import moment, {Moment} from "moment";
import {Sex} from "../../../demographic/model/Sex";
import MhaPatientAccess from "./MhaPatientAccess";
import MhaPatientAccessService from "../service/MhaPatientAccessService";
import MhaIntegration from "./MhaIntegration";

export default class MhaPatient
{
	protected _id: string;
	protected _firstName: string;
	protected _middleName: string;
	protected _lastName: string;
	protected _birthDate: Moment;
	protected _sex: Sex;

	protected _healthCareProvinceCode: Province;
	protected _healthNumber: string;
	protected _healthNumberVersion: string;

	protected _email: string;
	protected _cellPhone: string;
	protected _postalCode: string;
	protected _city: string;
	protected _address: string;
	protected _province: Province;
	protected _hasVoipToken: boolean;

	protected _linkStatus: LinkStatus;
	protected _canMessage: boolean;
	protected _demographicNo: string;

	// ==========================================================================
	// Public Methods
	// ==========================================================================

	/**
	 * get all patient access records for this patient across all MHA integrations
	 * @return promise that resolves to a list of access records
	 */
	public async getPatientAccessRecords(): Promise<MhaPatientAccess[]>
	{
		const accessService = new MhaPatientAccessService();
		return await accessService.getPatientAccesses(this.id);
	}

	/**
	 * get a specific access record for this patient
	 * @param integration - the integration to get the record from
	 * @return promise that resolves to an access record or null
	 */
	public async getPatientAccessRecord(integration: MhaIntegration): Promise<MhaPatientAccess>
	{
		const accessService = new MhaPatientAccessService();
		return await accessService.getPatientAccess(integration.id, this.id);
	}

	// ==========================================================================
	// Setters
	// ==========================================================================

	set id(value: string)
	{
		this._id = value;
	}

	set firstName(value: string)
	{
		this._firstName = value;
	}

	set middleName(value: string)
	{
		this._middleName = value;
	}

	set lastName(value: string)
	{
		this._lastName = value;
	}

	set birthDate(value: Moment)
	{
		this._birthDate = value;
	}

	set sex(value: Sex)
	{
		this._sex = value;
	}

	set healthCareProvinceCode(value: Province)
	{
		this._healthCareProvinceCode = value;
	}

	set healthNumber(value: string)
	{
		this._healthNumber = value;
	}

	set healthNumberVersion(value: string)
	{
		this._healthNumberVersion = value;
	}

	set email(value: string)
	{
		this._email = value;
	}

	set cellPhone(value: string)
	{
		this._cellPhone = value;
	}

	set postalCode(value: string)
	{
		this._postalCode = value;
	}

	set city(value: string)
	{
		this._city = value;
	}

	set address(value: string)
	{
		this._address = value;
	}

	set province(value: Province)
	{
		this._province = value;
	}

	set hasVoipToken(value: boolean)
	{
		this._hasVoipToken = value;
	}

	set linkStatus(value: LinkStatus)
	{
		this._linkStatus = value;
	}

	set canMessage(value: boolean)
	{
		this._canMessage = value;
	}

	set demographicNo(value: string)
	{
		this._demographicNo = value;
	}

	// ==========================================================================
	// Getters
	// ==========================================================================

	get id(): string
	{
		return this._id;
	}

	get firstName(): string
	{
		return this._firstName;
	}

	get middleName(): string
	{
		return this._middleName;
	}

	get lastName(): string
	{
		return this._lastName;
	}

	get displayName(): string
	{
		return `${this.firstName} ${this.lastName}`;
	}

	get birthDate(): Moment
	{
		return this._birthDate;
	}

	get age(): number
	{
		let now = moment();
		let age = moment.duration(now.diff(this.birthDate)).years();
		if (isNaN(age))
		{
			return null;
		}
		return age;
	}

	get sex(): Sex
	{
		return this._sex;
	}

	get healthCareProvinceCode(): Province
	{
		return this._healthCareProvinceCode;
	}

	get healthNumber(): string
	{
		return this._healthNumber;
	}

	get healthNumberVersion(): string
	{
		return this._healthNumberVersion;
	}

	get email(): string
	{
		return this._email;
	}

	get cellPhone(): string
	{
		return this._cellPhone;
	}

	get postalCode(): string
	{
		return this._postalCode;
	}

	get city(): string
	{
		return this._city;
	}

	get address(): string
	{
		return this._address;
	}

	get province(): Province
	{
		return this._province;
	}

	get fullAddress(): string
	{
		return `${this.address ? this._address : ''} ${this.city ? this.city : ''}, ${this.province ? this.province : ''}`
	}

	get hasVoipToken(): boolean
	{
		return this._hasVoipToken;
	}

	get linkStatus(): LinkStatus
	{
		return this._linkStatus;
	}

	get isConfirmed(): boolean
	{
		return this._linkStatus === LinkStatus.CONFIRMED || this._linkStatus === LinkStatus.VERIFIED;
	}

	get isVerified(): boolean
	{
		return this._linkStatus === LinkStatus.VERIFIED;
	}

	get isRejected(): boolean
	{
		return this._linkStatus === LinkStatus.CLINIC_REJECTED ||
			this._linkStatus === LinkStatus.PATIENT_REJECTED;
	}

	get canMessage(): boolean
	{
		return this._canMessage;
	}

	get demographicNo(): string
	{
		return this._demographicNo;
	}
}